### PR TITLE
Refactor speciation

### LINKED ISSUE #251: Refactoring of Speciation Move
The `SpeciationMove` and `ReactionData` needs refactoring. They are hard to read and to expand.

- [x] variable names should follow Faunus policies
- [x] code documentation is missing
- [x] Split into single task functions

### PR DESCRIPTION
**Pull request by mlund**
This is a refactoring commit that aims to fix #251. It is done in preparation for a guest coming in one week that will be working with speciation. There are still things to fix and improve, but that will be done along the way. When reviewing, please look for obvious overlooks and not major design changes (although such could be needed).

Squash and merge with edited log intended.

### CODE REVIEW

File: src/speciation.cpp [hunk 1]
```diff
@@ -8,250 +8,332 @@ namespace Move {
...
+
+    target.deactivate(target.begin(), target.end()); // deactivate whole group
+
+    Change::data change_data; // describes the change
+    change_data.internal = true;
+    change_data.index = &target - &spc.groups.front(); // index of moved group
+    change_data.all = true;                            // all atoms in group were moved
+    for (int i = 0; i < target.capacity(); i++) {      // list of all changed atom index
+        change_data.atoms.push_back(i);
+    }
```
Comment (rc83) [inline src/speciation.cpp]: Let’s be modern and use `std::iota`. ;-)

File: src/speciation.cpp [hunk 2]
```diff
@@ -8,250 +8,332 @@ namespace Move {
...
+
+    Change::data d;                          // describes the changed - used for energy evaluation
+    d.index = &target - &spc.groups.front(); // index* of moved group
+    d.all = true;                            // all atoms in group were moved
+    d.internal = true;
+    d.atoms.reserve(target.capacity());
+
+    for (size_t i = 0; i < target.capacity(); i++) { // list of changed atom index
+        d.atoms.push_back(i);
+    }
```
Comment (rc83) [inline src/speciation.cpp]: Let’s be modern and use `std::iota`. ;-)

File: src/molecule.h [hunk 3]
```diff
@@ -314,65 +314,96 @@ class NeighboursGenerator {
...
+    typedef std::map<int, int> Tmap; // key = molid; value = stoichiometic number
+    enum class Direction : char { LEFT = 0, RIGHT = 1 };
 
-    bool empty(bool forward) const;
+  private:
+    friend void from_json(const json &, ReactionData &);
+    friend void to_json(json &, const ReactionData &);
 
-    std::vector<int> participatingMolecules() const; //!< Returns molids of participating molecules
+    Direction direction = Direction::RIGHT;           //!< Direction of reaction
```
[the remark this answers is not in the document]
Comment (mlund) [reply src/molecule.h]: The finite resevoir is an incomplete feature and I have now labelled it so in the documentation

[the remark this answers is not in the document]
Comment (mlund) [reply src/molecule.h]: The view is a good idea. One way would be to have a base class and derived left and right objects. This functionality was previously achieved by letting all functions take a `bool forward` argument that was even more confusing. For now I'd like to leave it as is, as we'll need to come back to the speciation move again when finalising the "canonic" features.

File: src/molecule.cpp [hunk 4]
```diff
@@ -605,115 +605,117 @@ ParticleVector &Conformation::toParticleVector(ParticleVector &p) const {
     return p;
 }
 
-bool ReactionData::empty(bool forward) const {
-    if (forward)
-        if (canonic)
-            if (N_reservoir <= 0)
+bool ReactionData::empty() const {
+    if (direction == Direction::RIGHT) {
```
[the remark this answers is not in the document]
Comment (mlund) [reply src/molecule.cpp]: The system will not crash if there's either products, reactants, or both. The `empty()` function is called in `_move()` and will currently always return `false`

File: src/molecule.cpp [hunk 5]
```diff
@@ -605,115 +605,117 @@ ParticleVector &Conformation::toParticleVector(ParticleVector &p) const {
...
+        return {right_atoms, right_molecules};
+}
+void ReactionData::reverseDirection() {
+    if (direction == Direction::RIGHT)
+        setDirection(Direction::LEFT);
+    else
+        setDirection(Direction::RIGHT);
+}
 
 void from_json(const json &j, ReactionData &a) {
```
[the remark this answers is not in the document]
Comment (mlund) [reply src/molecule.cpp]: good idea

**Pull request merged**